File: backend/src/main/java/com/arudra/crm/entity/Workforce.java
package com.arudra.crm.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;

/**
 * Unified workforce header. Holds the ~90% of information common to every kind of workforce
 * (employees, contractors, and future freelancers/consultants/vendors). The type-specific fields
 * live in the extension tables — {@link Employee} and {@link Contractor} — each of which points back
 * here via its {@code workforce} reference.
 *
 * <p>{@code workforceType} is a plain String (matching the codebase convention that every type/status
 * field is a String, never a DB enum) so a new workforce type needs no schema change.
 */
@Getter
@Setter
@Entity
@Table(name = "workforce", indexes = {
    @Index(name = "idx_workforce_type", columnList = "workforce_type"),
    @Index(name = "idx_workforce_status", columnList = "status"),
    @Index(name = "idx_workforce_skill", columnList = "primary_skill")
})
public class Workforce extends BaseEntity {

    /** EMPLOYEE, CONTRACTOR, FREELANCER, CONSULTANT, VENDOR, TEMPORARY (extensible). */
    @Column(name = "workforce_type", nullable = false, length = 30)
    private String workforceType;

    // --- Basic information ----------------------------------------------------
    @Column(name = "full_name", nullable = false, length = 255)
    private String fullName;

    @Column(name = "profile_photo_url", length = 500)
    private String profilePhotoUrl;

    @Column(length = 20)
    private String mobile;

    @Column(length = 150)
    private String email;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Column(length = 20)
    private String gender;

    @Column(name = "address_line1", length = 255)
    private String addressLine1;

    @Column(name = "address_line2", length = 255)
    private String addressLine2;

    @Column(length = 100)
    private String city;

    @Column(length = 100)
    private String state;

    @Column(length = 100)
    private String country;

    @Column(length = 20)
    private String pincode;

    // --- Identity -------------------------------------------------------------
    @Column(name = "aadhaar_number", length = 20)
    private String aadhaarNumber;

    @Column(name = "pan_number", length = 20)
    private String panNumber;

    @Column(name = "driving_license", length = 50)
    private String drivingLicense;

    @Column(name = "passport_number", length = 50)
    private String passportNumber;

    // --- Emergency contact ----------------------------------------------------
    @Column(name = "emergency_contact_name", length = 100)
    private String emergencyContactName;

    @Column(name = "emergency_relationship", length = 50)
    private String emergencyRelationship;

    @Column(name = "emergency_phone", length = 20)
    private String emergencyPhone;

    // --- Skills ---------------------------------------------------------------
    @Column(name = "primary_skill", length = 100)
    private String primarySkill;

    /** Comma-separated secondary skills. */
    @Column(name = "secondary_skills", columnDefinition = "TEXT")
    private String secondarySkills;

    @Column(name = "experience_years")
    private Integer experienceYears;

    @Column(columnDefinition = "TEXT")
    private String certifications;

    // --- Project info ---------------------------------------------------------
    @Column(name = "available_from")
    private LocalDate availableFrom;

    /** AVAILABLE, BUSY, ON_LEAVE, INACTIVE. */
    @Column(nullable = false, length = 30)
    private String status = "AVAILABLE";

    @Column(columnDefinition = "TEXT")
    private String notes;
}

File: frontend/src/types/workforce.ts
// Unified workforce resource model — the Project module treats employees and contractors as one
// pool of assignable resources. Mirrors backend com.arudra.crm.entity.ResourceType and
// dto/workforce/WorkforceResourceView.

export type ResourceType =
  | 'EMPLOYEE'
  | 'CONTRACTOR'
  | 'FREELANCER'
  | 'SUBCONTRACTOR'
  | 'CONSULTANT'
  | 'VENDOR';

export interface WorkforceResourceView {
  resourceType: ResourceType;
  resourceId: number;
  name: string;
  subtitle?: string;
  contact?: string;
}

export interface WorkforceProductivityRow {
  resourceType: ResourceType;
  resourceId: number;
  name: string;
  assigned: number;
  completed: number;
  overdue: number;
  completionRate: number;
}

export interface WorkforceDashboard {
  totalAssignments: number;
  completedAssignments: number;
  inProgressAssignments: number;
  overdueAssignments: number;
  employeeAssignments: number;
  contractorAssignments: number;
  distinctResources: number;
  productivity: WorkforceProductivityRow[];
}

export const RESOURCE_TYPE_LABELS: Record<string, string> = {
  EMPLOYEE: 'Employee',
  CONTRACTOR: 'Contractor',
  FREELANCER: 'Freelancer',
  SUBCONTRACTOR: 'Subcontractor',
  CONSULTANT: 'Consultant',
  VENDOR: 'Vendor',
};

export const RESOURCE_TYPE_STYLES: Record<string, string> = {
  EMPLOYEE: 'bg-emerald-100 text-emerald-700',
  CONTRACTOR: 'bg-amber-100 text-amber-700',
  FREELANCER: 'bg-teal-100 text-teal-700',
  SUBCONTRACTOR: 'bg-orange-100 text-orange-700',
  CONSULTANT: 'bg-purple-100 text-purple-700',
  VENDOR: 'bg-cyan-100 text-cyan-700',
};

// ---------------------------------------------------------------------------
// Unified Workforce MASTER model (single "Add Workforce" flow, one directory, one profile).
// Mirrors backend dto/workforce/{WorkforceRequest,WorkforceListRow,WorkforceDetailView} and the
// Workforce entity. Kept in the same file as the assignment-side types above.
// ---------------------------------------------------------------------------

export type WorkforceStatus = 'AVAILABLE' | 'BUSY' | 'ON_LEAVE' | 'INACTIVE';

export const WORKFORCE_STATUSES: WorkforceStatus[] = ['AVAILABLE', 'BUSY', 'ON_LEAVE', 'INACTIVE'];

export const WORKFORCE_STATUS_TONE: Record<string, string> = {
  AVAILABLE: 'bg-emerald-100 text-emerald-700',
  BUSY: 'bg-amber-100 text-amber-700',
  ON_LEAVE: 'bg-emerald-100 text-emerald-700',
  INACTIVE: 'bg-slate-100 text-slate-600',
};

export interface WorkforceListRow {
  id: number;
  workforceType: ResourceType;
  fullName: string;
  primarySkill?: string;
  status: WorkforceStatus;
  activeProjects: number;
  mobile?: string;
  email?: string;
  profilePhotoUrl?: string;
  employeeId?: number;
  department?: string;
  contractorId?: number;
  companyName?: string;
}

/** The workforce header entity (shared fields). */
export interface Workforce {
  id: number;
  workforceType: ResourceType;
  fullName: string;
  profilePhotoUrl?: string;
  mobile?: string;
  email?: string;
  dateOfBirth?: string;
  gender?: string;
  addressLine1?: string;
  addressLine2?: string;
  city?: string;
  state?: string;
  country?: string;
  pincode?: string;
  aadhaarNumber?: string;
  panNumber?: string;
  drivingLicense?: string;
  passportNumber?: string;
  emergencyContactName?: string;
  emergencyRelationship?: string;
  emergencyPhone?: string;
  primarySkill?: string;
  secondarySkills?: string;
  experienceYears?: number;
  certifications?: string;
  availableFrom?: string;
  status: WorkforceStatus;
  notes?: string;
}

export interface WorkforceDocument {
  id: number;
  docType: string;
  fileName: string;
  fileUrl: string;
  notes?: string;
}

export interface WorkforceDetail {
  workforce: Workforce;
  employee?: any;    // Employee extension (payroll/department/designation)
  contractor?: any;  // Contractor extension (company/GST/rates/compliance)
  documents: WorkforceDocument[];
  activeProjects: number;
  attendanceRequired: boolean;
}

export interface EmployeeDetailsInput {
  employeeCode?: string;
  departmentId?: number;
  designation?: string;
  joiningDate?: string;
  salary?: number;
  salaryType?: string;
  shift?: string;
  attendanceRequired?: boolean;
  leavePolicy?: string;
  payrollEnabled?: boolean;
  pfNumber?: string;
  esiNumber?: string;
  bankAccount?: string;
  ifsc?: string;
  uan?: string;
}

export interface ContractorDetailsInput {
  companyName?: string;
  contractorCode?: string;
  contactPerson?: string;
  gstNumber?: string;
  contractStartDate?: string;
  contractEndDate?: string;
  labourRate?: number;
  paymentTerms?: string;
  agreementNumber?: string;
  serviceCategories?: string;
  tdsApplicable?: boolean;
  insuranceDetails?: string;
}

export interface WorkforceRequest {
  workforceType: ResourceType;
  fullName?: string;
  profilePhotoUrl?: string;
  mobile?: string;
  email?: string;
  dateOfBirth?: string;
  gender?: string;
  addressLine1?: string;
  addressLine2?: string;
  city?: string;
  state?: string;
  country?: string;
  pincode?: string;
  aadhaarNumber?: string;
  panNumber?: string;
  drivingLicense?: string;
  passportNumber?: string;
  emergencyContactName?: string;
  emergencyRelationship?: string;
  emergencyPhone?: string;
  primarySkill?: string;
  secondarySkills?: string;
  experienceYears?: number;
  certifications?: string;
  availableFrom?: string;
  status?: WorkforceStatus;
  notes?: string;
  employee?: EmployeeDetailsInput;
  contractor?: ContractorDetailsInput;
}

export interface WorkforceTypeMeta {
  value: ResourceType;
  label: string;
  creatable: boolean;
}

export interface WorkforceMeta {
  types: WorkforceTypeMeta[];
  statuses: string[];
  skills: string[];
  salaryTypes: string[];
  departments: { id: number; name: string }[];
}
